docs/proposals/user-namespace.md

```diff
@@ -0,0 +1,226 @@
+configuration.
+
+### Known Issues
+
+These are the known issues (as of writing) that will be hit when
+enabling user namespace remapping in a Kubernetes cluster.
+
+1. Negative oomScoreAdj values fail.  In this case the pause container
+cannot be started and shows an error `write /proc/self/oom_score_adj: permission denied`
+in the daemon log.  Failing in docker 1.10, working in 1.12
+1. Cannot share the host or a container's network namespace when user
+namespaces are enabled.  [Fixed in 1.11](https://github.com/docker/docker/pull/21383).
+1. Permission denied mounting.  Occurs due to ownership permissions
+in the `/var/lib/kubelet/pods` directory when trying to mount the
+termination logs and secrets.  A workaround (not a solution) is to
+change the remapped GID to the root group.
+
+## Design
+
+### Constraints
+
+* If containers share any namespace they must share the user namespace.
+This is because [capabilities are checked against the user namespace](http://man7.org/linux/man-pages/man7/user_namespaces.7.html).
+This means that if any namespace is set to `host` the pod must
+also use the host user namespace.  Since containers in Kubernetes
+always share a network namespace they must also share the user
+namespace.
+* Currently you may only specify a single user namespace remap value
+to the docker daemon.  This means you cannot ensure namespaces are
+using different UIDs and GIDs for remapping, they will all be remapped
+to the same value(s).
+* [Known restrictions when running in a remap environment](https://docs.docker.com/engine/reference/commandline/dockerd/#user-namespace-known-restrictions)
+
+### Open Questions
+
+1.  Do we need to expose a `HostUserNamespace` option on the `SecurityContext`
+or can we provide defaulting mechanisms that do the right thing?
+1.  If we are not defaulting and we are aware of the remap environment
+should we validate in the API server for misconfiguration or allow
+this to fail in the container runtime?
+1.  Is there a valid use case for running some nodes with remapping
+turned on and some without?
+
+### Option 1: System Defaulting
+
+In this option, there is no `HostUserNamespace` flag exposed in the API.
+This requires that the Kubelet be aware of the remapped environment and
+provide defaulting for the host user namespace settings of a container
+runtime when launching the containers.  This work has been
+[prototyped](https://github.com/kubernetes/kubernetes/pull/31169).
+
+It is debatable whether or not the kubelet should provide defaulting based on
+this knowledge since it may mean that a container is run with a
+configuration that is not what the pod author expected.  However, if
+policy can check that the defaulted configuration is valid for the
+user it may be valuable and avoid frustrations with runtime errors.
+
+Pros:
+
+1. reduces the need for Linux specific knowledge
+1. system does the right thing
+
+Cons:
+
+1. requires extra flags to make the system aware of the remapped
```

Comment:
It'd be nice to be able to detect this, instead of adding yet another flag to kubelet.